app/lib/actions/katello/event_queue/monitor.rb

```diff
@@ -0,0 +1,107 @@
+          fields! backtrace: String, message: String, kind: String
+        end
+
+        Ready = Algebrick.atom
+        Close = Algebrick.atom
+
+        attr_reader :triggered_action
+
+        def self.ensure_running(world = ForemanTasks.dynflow.world)
+          world.coordinator.acquire(RunOnceCoordinatorLock.new(world)) do
+            unless ForemanTasks::Task::DynflowTask.for_action(self).running.any?
+              @triggered_action = ForemanTasks.trigger(self)
+            end
+          end
+        rescue Dynflow::Coordinator::LockError
+          return false
+        end
+
+        def plan
+          # Make sure we don't have two concurrent listening services competing
+          if already_running?
+            fail "Action #{self.class.name} is already active"
+          end
+          plan_self
+        end
+
+        def run(event = nil)
+          match(event,
+                (on nil do
+                  initialize_service
+                end),
+                (on Ready do
+                  listen_for_events
+                end),
+                (on Event do
+                   act_on_event(event)
+                 end),
+                (on Close | Dynflow::Action::Cancellable::Cancel do
+                   close_service
+                 end),
+                (on Fatal do
+                   close_service_with_error(event)
+                 end),
+                (on Dynflow::Action::Skip do
+                   # do nothing, just skip
+                 end))
+        rescue => e
+          action_logger.error(e.message)
+          close_service
+          error!(e)
+        end
+
+        def close_service_with_error(event)
+          error = Exception.new("#{event.kind}: #{event.message}")
+          error.set_backtrace([event.backtrace])
+          error!(error)
+        ensure
+          close_service
+        end
+
+        def close_service
+          PollerThread.instance.close
+        end
+
+        def initialize_service
+          ::Katello::EventQueue.reset_in_progress
+          PollerThread.initialize(world.logger)
+          suspend do |suspended_action|
+            SuspendedAction.new(suspended_action).notify_ready
+          end
+        end
+
+        def listen_for_events
+          output[:error] = nil
+          suspend do |suspended_action|
+            PollerThread.instance.poll_for_events(SuspendedAction.new(suspended_action))
+          end
+        end
+
+        def act_on_event(event)
+          begin
+            User.as_anonymous_admin do
+              output[:last_event] = "#{event.event_type} - #{event.object_id}"
+              ::Katello::EventQueue.event_class(event.event_type).new(event.object_id).run
+            end
+          rescue => e
+            output[:last_error] = e.message
+            error!(e)
+            raise e
+          ensure
+            ::Katello::EventQueue.clear_events(event.event_type, event.object_id, event.created_at)
+          end
+          suspend
```

Comment:
Should we put this in the `begin` block since it would only get run if that is successful? Here it sorta looks like it should run always but due to the rescue and ensure that isn't the case.